mrmustard/physics/ansatze.py

```diff
@@ -0,0 +1,351 @@
+
+    def simplify_v2(self) -> None:
+        r"""A different implementation that orders the batch dimension first."""
+        if self._simplified:
+            return
+        self._order_batch()
+        to_keep = [d0 := 0]
+        mat, vec = self.mat[d0], self.vec[d0]
+        for d in range(1, self.batch_size):
+            if np.allclose(mat, self.mat[d]) and np.allclose(vec, self.vec[d]):
+                self.array = math.update_add_tensor(self.array, [d0], [self.array[d]])
+            else:
+                to_keep.append(d)
+                d0 = d
+                mat, vec = self.mat[d0], self.vec[d0]
+        self.mat = math.gather(self.mat, to_keep, axis=0)
+        self.vec = math.gather(self.vec, to_keep, axis=0)
+        self.array = math.gather(self.array, to_keep, axis=0)
+        self._simplified = True
+
+    def _order_batch(self):
+        generators = [
+            itertools.chain(
+                math.asnumpy(self.vec[i]).flat,
+                math.asnumpy(self.mat[i]).flat,
+                math.asnumpy(self.array[i]).flat,
+            )
+            for i in range(self.batch_size)
+        ]
+        sorted_indices = argsort_gen(generators)
+        self.mat = math.gather(self.mat, sorted_indices, axis=0)
+        self.vec = math.gather(self.vec, sorted_indices, axis=0)
+        self.array = math.gather(self.array, sorted_indices, axis=0)
+
+
+class PolyExpAnsatz(PolyExpBase):
+    """
+    Represents the ansatz function:
+
+        F(z) = sum_i poly_i(z) exp(z^T A_i z / 2 + z^T b_i),
+
+    where each poly_i is a polynomial in z that can be expressed as:
+
+        poly_i(z) = sum_k c^(i)_k z^k,
+
+    with k being a multi-index. The batch of arrays c^(i) are not just array values but can be polynomials
+    of varying order, defined by the terms arr_k z^k for each i. The matrices A_i and vectors b_i
+    are parameters of the exponential terms in the ansatz, and z is a vector of variables.
+
+    Attributes:
+        A (Batch[Matrix]): The list of square matrices A_i
+        b (Batch[Vector]): The list of vectors b_i
+        c (Batch[Tensor]): The array of coefficients for the polynomial terms in the ansatz.
+
+    Example:
+        >>> A = [np.array([[1.0, 0.0], [0.0, 1.0]])]
+        >>> b = [np.array([1.0, 1.0])]
+        >>> c = [np.array(1.0)]
+        >>> F = PolyExpAnsatz(A, b, c)
+        >>> z = np.array([1.0, 2.0])
+        >>> print(F(z))  # prints the value of F at z
+    """
+
+    def __init__(self, A: Batch[Matrix], b: Batch[Vector], c: Batch[Tensor], name: str = ""):
+        self.name = name
+        super().__init__(mat=A, vec=b, array=c)
+
+    @property
+    def A(self) -> Batch[ComplexMatrix]:
+        return self.mat
+
+    @property
+    def b(self) -> Batch[ComplexVector]:
+        return self.vec
+
+    @property
+    def c(self) -> Batch[ComplexTensor]:
+        return self.array
+
+    @property
+    def degree(self) -> int:
+        return self.array.shape[-1] - 1
+
+    def plot(
```

Comment:
I tried this:

`from mrmustard.physics.representations import Bargmann`
`import numpy as np`
`a = np.zeros((2, 2))`
`b = np.zeros((2,))`
`vacuum = Bargmann(a, b)`
`vacuum.plot()`

The title is weird (It has `F{}..` with empty brakets). Also, it prints `(<Figure size 640x480 with 1 Axes>,
 <AxesSubplot:title={'center':'F_{}(z)exp(-|z|^2)'}, xlabel='$Re(z)$', ylabel='$Im(z)$'>)` at the bottom

Reply:
yep the labels need some fixing. The statements you see are because the method returns (fig,ax) in case you want to modify/customize the plot futher

Reply:
I'll return ax only